### PR TITLE
Increase test coverage of upgrade.controller.js

### PR DESCRIPTION
**Pull request by Itxaka**
My first time adding this type of tests where you need to mock some services so not sure if everything is ok, so please check carefully. Thanks!

Still missing tests for refeshStepsList, they will come tomorrow in a new commit

Before:

```
=============================== Coverage summary ===============================
Statements   : 97.26% ( 1279/1315 )
Branches     : 81.48% ( 44/54 )
Functions    : 97.74% ( 519/531 )
Lines        : 97.26% ( 1279/1315 )
================================================================================
```

After

```
=============================== Coverage summary ===============================
Statements   : 99.04% ( 1347/1360 )
Branches     : 87.04% ( 47/54 )
Functions    : 97.98% ( 533/544 )
Lines        : 99.04% ( 1347/1360 )
================================================================================
```

### CODE REVIEW

File: assets/app/features/upgrade/upgrade.controller.spec.js [hunk 1]
```diff
@@ -1,7 +1,105 @@
...
+        });
+
+        describe('nextStep function', function () {
+            it('should be defined', function () {
+                should.exist(controller.steps.nextStep);
+                expect(controller.steps.nextStep).toEqual(jasmine.any(Function));
+            });
+
+            it('should change step when called', function () {
+                expect(controller.steps.activeStep.id).toEqual(0);
```
Comment (skazi0) [inline assets/app/features/upgrade/upgrade.controller.spec.js:65]: The fact that after creation activeStep points to `0` is used in multiple tests but it's actually part of controller's behavior which deserves its own test before we assume it's valid as a pre-condition in other tests.

**Review by skazi0 — COMMENTED** [review]
Looks good so far. One comment inline.

File: assets/app/features/upgrade/upgrade.controller.spec.js (hunk 1 above)
Comment (Itxaka) [reply assets/app/features/upgrade/upgrade.controller.spec.js:65]: Good point!

**Comment by Itxaka** [comment]
Updated and ready for review. upgrade.controller.js is now at 100% coverage :tada: 

```
=============================== Coverage summary ===============================
Statements   : 99.34% ( 1354/1363 )
Branches     : 88.89% ( 48/54 )
Functions    : 97.98% ( 533/544 )
Lines        : 99.34% ( 1354/1363 )
================================================================================
```

File: assets/app/features/upgrade/upgrade.controller.spec.js [hunk 2]
```diff
@@ -1,7 +1,108 @@
...
+            });
+
+            it('on first step should return false', function () {
+                assert.isFalse(controller.steps.isLastStep());
+            });
+
+            it('after moving to last step should return true', function () {
+                expect(controller.steps.activeStep.id).toEqual(0);
+                for (var i = 0; i < (stepsList.length - 1); i++) {
+                    controller.steps.nextStep();
```
Comment (santiph) [inline assets/app/features/upgrade/upgrade.controller.spec.js]: `nextStep()` is out of the scope of `isLastStep()` unit test.
This should be achieved through a configuration phase:

```
beforeEach(function() {
    $state.current  = _.last(stepsList);
});
```

File: assets/app/features/upgrade/upgrade.controller.spec.js [hunk 3]
```diff
@@ -1,7 +1,108 @@
...
+            it('should be defined', function () {
+                should.exist(controller.steps.nextStep);
+                expect(controller.steps.nextStep).toEqual(jasmine.any(Function));
+            });
+
+            it('should change step when called', function () {
+                expect(controller.steps.activeStep.id).toEqual(0);
+                controller.steps.nextStep();
+                expect(controller.steps.activeStep.id).toEqual(1);
+
```
Comment (santiph) [inline assets/app/features/upgrade/upgrade.controller.spec.js:68]: `$state.go(vm.steps.activeStep.state);` call needs also to be tested, but won't be possible with bard.
Please take a look at https://github.com/crowbar/crowbar-ui/blob/master/assets/app/features/upgrade/controllers/upgrade-landing.controller.spec.js#L251-L305

File: assets/app/features/upgrade/upgrade.controller.spec.js [hunk 4]
```diff
@@ -1,7 +1,108 @@
...
+            });
+
+            it('should not change step if we are on the last step', function () {
+                expect(controller.steps.activeStep.id).toEqual(0);
+                for (var i = 0; i < (stepsList.length - 1); i++) {
```
Comment (santiph) [inline assets/app/features/upgrade/upgrade.controller.spec.js:73]: Same comments as https://github.com/crowbar/crowbar-ui/pull/74/files#r84088522

File: assets/app/features/upgrade/upgrade.controller.spec.js (hunk 3 above)
Comment (Itxaka) [reply assets/app/features/upgrade/upgrade.controller.spec.js:68]: Do we really need to test the $state.go functionality? isnt that part of the $stateProvider test suite? Seems like out of scope for our tests to check for that.

File: assets/app/features/upgrade/upgrade.controller.spec.js (hunk 2 above)
Comment (Itxaka) [reply assets/app/features/upgrade/upgrade.controller.spec.js]: done!

File: assets/app/features/upgrade/upgrade.controller.spec.js (hunk 4 above)
Comment (Itxaka) [reply assets/app/features/upgrade/upgrade.controller.spec.js:73]: This is actually testing nextStep() to see if it goes over the last step when its at the end so I think this should stay (could be more elegant than a for, that is true)

File: assets/app/features/upgrade/upgrade.controller.spec.js (hunk 3 above)
Comment (santiph) [reply assets/app/features/upgrade/upgrade.controller.spec.js:68]: We don't need to validate its inner functionality, but we do care about the `$state.go()` call and we should to assert it's being called with a particular parameter.

**Comment by Itxaka** [comment]
It may make more sense to merge any tests from this PR into https://github.com/crowbar/crowbar-ui/issues/71 which changes the upgrade.controller.js as to have full coverage without having to accomodate this? @santiph ?

**Comment by Itxaka** [comment]
was having a lot of issues to rebase + rebuild the tests to be in sync with changes in master so I just redid them fresh, which is better.
Closing this in favor of https://github.com/crowbar/crowbar-ui/pull/88

**Comment by skazi0** [comment]
Why not `reset` + `rebase` + `push --force` to the same branch to avoid additional PRs?

**Comment by Itxaka** [comment]
@skazi0 almost no code from this pr made it final + a lot of time rebasing + broke somthing during rebase and had to redo it. Too much time spent trying to fix it tbh

**Comment by skazi0** [comment]
@Itxaka I got this from your comment. I think you could still just drop all of your changes with `git reset` and then cleanly rebase on top of current master.

**Comment by Itxaka** [comment]
Umm, indeed good point @skazi0 

I guess that I just got too entangled with the changes and blocked and could not see it :/

**Pull request closed without merging**